

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>indra.databases.uniprot_client &mdash; INDRA 1.10.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="INDRA 1.10.0 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> INDRA
          

          
          </a>

          
            
            
              <div class="version">
                1.10
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license.html">License and funding</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting started with INDRA</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/index.html">INDRA modules reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../rest_api.html">REST API</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">INDRA</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>indra.databases.uniprot_client</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for indra.databases.uniprot_client</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">absolute_import</span><span class="p">,</span> <span class="n">print_function</span><span class="p">,</span> <span class="n">unicode_literals</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="k">import</span> <span class="nb">dict</span><span class="p">,</span> <span class="nb">str</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">rdflib</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">from</span> <span class="nn">xml.etree</span> <span class="k">import</span> <span class="n">ElementTree</span>
<span class="k">try</span><span class="p">:</span>
    <span class="c1"># Python 3</span>
    <span class="kn">from</span> <span class="nn">functools</span> <span class="k">import</span> <span class="n">lru_cache</span>
    <span class="kn">from</span> <span class="nn">urllib.error</span> <span class="k">import</span> <span class="n">HTTPError</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="c1"># Python 2</span>
    <span class="kn">from</span> <span class="nn">functools32</span> <span class="k">import</span> <span class="n">lru_cache</span>
    <span class="kn">from</span> <span class="nn">urllib2</span> <span class="k">import</span> <span class="n">HTTPError</span>
<span class="kn">from</span> <span class="nn">indra.util</span> <span class="k">import</span> <span class="n">read_unicode_csv</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;uniprot&#39;</span><span class="p">)</span>

<span class="n">uniprot_url</span> <span class="o">=</span> <span class="s1">&#39;http://www.uniprot.org/uniprot/&#39;</span>

<span class="n">rdf_prefixes</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    PREFIX up: &lt;http://purl.uniprot.org/core/&gt;</span>
<span class="s2">    PREFIX db: &lt;http://purl.uniprot.org/database/&gt;</span>
<span class="s2">    PREFIX faldo: &lt;http://biohackathon.org/resource/faldo#&gt;</span>
<span class="s2">    PREFIX rdf: &lt;http://www.w3.org/1999/02/22-rdf-syntax-ns#&gt;</span>
<span class="s2">    PREFIX rdfs: &lt;http://www.w3.org/2000/01/rdf-schema#&gt; &quot;&quot;&quot;</span>


<div class="viewcode-block" id="query_protein"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.query_protein">[docs]</a><span class="nd">@lru_cache</span><span class="p">(</span><span class="n">maxsize</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt entry as an RDF graph for the given UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID to be queried.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    g : rdflib.Graph</span>
<span class="sd">        The RDF graph corresponding to the UniProt entry.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Try looking up a primary ID if the given one</span>
    <span class="c1"># is a secondary ID</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">prim_ids</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_sec</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span>
        <span class="n">protein_id</span> <span class="o">=</span> <span class="n">prim_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="n">url</span> <span class="o">=</span> <span class="n">uniprot_url</span> <span class="o">+</span> <span class="n">protein_id</span> <span class="o">+</span> <span class="s1">&#39;.rdf&#39;</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">rdflib</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">g</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">HTTPError</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Could not find protein with id </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">protein_id</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">except</span> <span class="n">rdflib</span><span class="o">.</span><span class="n">exceptions</span><span class="o">.</span><span class="n">ParserError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;Could not parse RDF at </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">url</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># Check if the entry has been replaced by a new entry</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?res2</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?res1 up:replacedBy ?res2 .</span>
<span class="s2">            }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">term</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">replaced_by_id</span> <span class="o">=</span> <span class="n">term</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">replaced_by_id</span><span class="p">)</span></div>
    <span class="k">return</span> <span class="n">g</span>

<div class="viewcode-block" id="is_secondary"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.is_secondary">[docs]</a><span class="k">def</span> <span class="nf">is_secondary</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the UniProt ID corresponds to a secondary accession.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        The UniProt ID to check.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if it is a secondary accessing entry, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">entry</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_sec</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">entry</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span></div>
    <span class="k">return</span> <span class="kc">True</span>

<div class="viewcode-block" id="get_primary_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_primary_id">[docs]</a><span class="k">def</span> <span class="nf">get_primary_id</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a primary entry corresponding to the UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        The UniProt ID to map to primary.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    primary_id : str</span>
<span class="sd">        If the given ID is primary, it is returned as is. Othwewise the primary</span>
<span class="sd">        IDs are looked up. If there are multiple primary IDs then the first</span>
<span class="sd">        human one is returned. If there are no human primary IDs then the</span>
<span class="sd">        first primary found is returned.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">primaries</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_sec</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">primaries</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">primaries</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;More than 1 primary ID for </span><span class="si">%s</span><span class="s1">.&#39;</span> <span class="o">%</span> <span class="n">protein_id</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">primary</span> <span class="ow">in</span> <span class="n">primaries</span><span class="p">:</span>
                <span class="c1"># Often secondary IDs were broken into multiple primary IDs</span>
                <span class="c1"># for different organisms. In this case we return the human</span>
                <span class="c1"># one if it exists.</span>
                <span class="k">if</span> <span class="n">is_human</span><span class="p">(</span><span class="n">primary</span><span class="p">):</span>
                    <span class="k">return</span> <span class="n">primary</span>
        <span class="c1"># If we haven&#39;t returned anything then we just return the</span>
        <span class="c1"># first primary id</span>
        <span class="k">return</span> <span class="n">primaries</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># If there is not secondary entry the we assume this is a primary entry</span></div>
    <span class="k">return</span> <span class="n">protein_id</span>

<div class="viewcode-block" id="get_family_members"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_family_members">[docs]</a><span class="k">def</span> <span class="nf">get_family_members</span><span class="p">(</span><span class="n">family_name</span><span class="p">,</span> <span class="n">human_only</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the HGNC gene symbols which are the members of a given family.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    family_name : str</span>
<span class="sd">        Family name to be queried.</span>
<span class="sd">    human_only : bool</span>
<span class="sd">        If True, only human proteins in the family will be returned.</span>
<span class="sd">        Default: True</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    gene_names : list</span>
<span class="sd">        The HGNC gene symbols corresponding to the given family.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;query&#39;</span><span class="p">:</span> <span class="s1">&#39;family:</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">family_name</span><span class="p">,</span> 
            <span class="s1">&#39;format&#39;</span><span class="p">:</span> <span class="s1">&#39;list&#39;</span><span class="p">}</span>
    <span class="k">if</span> <span class="n">human_only</span><span class="p">:</span>
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;fil&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;organism:human&#39;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">uniprot_url</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">res</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">res</span><span class="o">.</span><span class="n">text</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="c1"># res.text gets us the Unicode</span>
    <span class="n">html</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">text</span>
    <span class="n">protein_list</span> <span class="o">=</span> <span class="n">html</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">protein_list</span><span class="p">:</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="n">get_gene_name</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="n">gene_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span></div>
    <span class="k">return</span> <span class="n">gene_names</span>

<div class="viewcode-block" id="get_mnemonic"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_mnemonic">[docs]</a><span class="k">def</span> <span class="nf">get_mnemonic</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">web_fallback</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt mnemonic for the given UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID to be mapped.</span>
<span class="sd">    web_fallback : Optional[bool]</span>
<span class="sd">        If True and the offline lookup fails, the UniProt web service</span>
<span class="sd">        is used to do the query.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mnemonic : str</span>
<span class="sd">        The UniProt mnemonic corresponding to the given Uniprot ID.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">mnemonic</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_mnemonic</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">mnemonic</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">web_fallback</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">g</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?mnemonic</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?r up:mnemonic ?mnemonic .</span>
<span class="s2">        }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">mnemonic</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">toPython</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">mnemonic</span>
    <span class="k">else</span><span class="p">:</span></div>
        <span class="k">return</span> <span class="kc">None</span>

<div class="viewcode-block" id="get_id_from_mnemonic"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_id_from_mnemonic">[docs]</a><span class="k">def</span> <span class="nf">get_id_from_mnemonic</span><span class="p">(</span><span class="n">uniprot_mnemonic</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt ID for the given UniProt mnemonic.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    uniprot_mnemonic : str</span>
<span class="sd">        UniProt mnemonic to be mapped.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    uniprot_id : str</span>
<span class="sd">        The UniProt ID corresponding to the given Uniprot mnemonic.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">uniprot_id</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_mnemonic_reverse</span><span class="p">[</span><span class="n">uniprot_mnemonic</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">uniprot_id</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span></div>
        <span class="k">return</span> <span class="kc">None</span>

<div class="viewcode-block" id="get_gene_name"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_gene_name">[docs]</a><span class="k">def</span> <span class="nf">get_gene_name</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">web_fallback</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the gene name for the given UniProt ID.</span>

<span class="sd">    This is an alternative to get_hgnc_name and is useful when</span>
<span class="sd">    HGNC name is not availabe (for instance, when the organism</span>
<span class="sd">    is not homo sapiens).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID to be mapped.</span>
<span class="sd">    web_fallback : Optional[bool]</span>
<span class="sd">        If True and the offline lookup fails, the UniProt web service</span>
<span class="sd">        is used to do the query.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    gene_name : str</span>
<span class="sd">        The gene name corresponding to the given Uniprot ID.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_id</span> <span class="o">=</span> <span class="n">get_primary_id</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_gene_name</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span>
        <span class="c1"># There are cases when the entry is in the resource</span>
        <span class="c1"># table but the gene name is empty. Often this gene</span>
        <span class="c1"># name is actually available in the web service RDF</span>
        <span class="c1"># so here we return only if the gene name is not None</span>
        <span class="c1"># and not empty string.</span>
        <span class="k">if</span> <span class="n">gene_name</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">gene_name</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">web_fallback</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">g</span> <span class="o">=</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">g</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?name</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?gene a up:Gene .</span>
<span class="s2">            ?gene skos:prefLabel ?name .</span>
<span class="s2">            }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">toPython</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_name</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="n">gene_name</span></div>
    <span class="k">return</span> <span class="kc">None</span>


<div class="viewcode-block" id="get_gene_synonyms"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_gene_synonyms">[docs]</a><span class="k">def</span> <span class="nf">get_gene_synonyms</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a list of synonyms for the gene corresponding to a protein.</span>

<span class="sd">    Note that synonyms here also include the official gene name as</span>
<span class="sd">    returned by get_gene_name.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        The UniProt ID of the protein to query</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    synonyms : list[str]</span>
<span class="sd">        The list of synonyms of the gene corresponding to the protein</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">g</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?name</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?gene skos:altLabel | skos:prefLabel ?name .</span>
<span class="s2">            }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">toPython</span><span class="p">()</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">]</span></div>
    <span class="k">return</span> <span class="kc">None</span>


<div class="viewcode-block" id="get_protein_synonyms"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_protein_synonyms">[docs]</a><span class="k">def</span> <span class="nf">get_protein_synonyms</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a list of synonyms for a protein.</span>

<span class="sd">    Note that this function returns protein synonyms as provided by UniProt.</span>
<span class="sd">    The get_gene_synonym returns synonyms given for the gene corresponding</span>
<span class="sd">    to the protein, and get_synonyms returns both.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        The UniProt ID of the protein to query</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    synonyms : list[str]</span>
<span class="sd">        The list of synonyms of the protein</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">g</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?name</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?gene :fullName | :shortName ?name .</span>
<span class="s2">            }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">toPython</span><span class="p">()</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">]</span></div>
    <span class="k">return</span> <span class="kc">None</span>


<div class="viewcode-block" id="get_synonyms"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_synonyms">[docs]</a><span class="k">def</span> <span class="nf">get_synonyms</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return synonyms for a protein and its associated gene.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        The UniProt ID of the protein to query</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    synonyms : list[str]</span>
<span class="sd">        The list of synonyms of the protein and its associated gene.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ret</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_syms</span> <span class="o">=</span> <span class="n">get_gene_synonyms</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gene_syms</span><span class="p">:</span>
        <span class="n">ret</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">gene_syms</span><span class="p">)</span>
    <span class="n">prot_syms</span> <span class="o">=</span> <span class="n">get_protein_synonyms</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">prot_syms</span><span class="p">:</span>
        <span class="n">ret</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">prot_syms</span><span class="p">)</span></div>
    <span class="k">return</span> <span class="n">ret</span>


<span class="nd">@lru_cache</span><span class="p">(</span><span class="n">maxsize</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">get_sequence</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">prim_ids</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_sec</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span>
        <span class="n">protein_id</span> <span class="o">=</span> <span class="n">prim_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="n">url</span> <span class="o">=</span> <span class="n">uniprot_url</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">.fasta&#39;</span> <span class="o">%</span> <span class="n">protein_id</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">res</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Could not find sequence for protein </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">protein_id</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="c1"># res.text is Unicode</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">lines</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq</span>


<span class="k">def</span> <span class="nf">get_modifications</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">g</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?beg_pos ?comment</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?mod_res a up:Modified_Residue_Annotation .</span>
<span class="s2">            ?mod_res rdfs:comment ?comment .</span>
<span class="s2">            ?mod_res up:range ?range .</span>
<span class="s2">            ?range faldo:begin ?beg .</span>
<span class="s2">            ?range faldo:end ?end .</span>
<span class="s2">            ?beg a faldo:ExactPosition .</span>
<span class="s2">            ?beg faldo:position ?beg_pos .</span>
<span class="s2">            FILTER (?beg = ?end)</span>
<span class="s2">            }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="n">mods</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">mod_pos</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>
        <span class="c1"># &quot;Phosphothreonine; by autocatalysis&quot;</span>
        <span class="c1"># &quot;Phosphothreonine; by MAP2K1 and MAP2K2&quot;</span>
        <span class="c1"># TODO: take into account the comment after the ;?</span>
        <span class="n">mod_res</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">value</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">mods</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mod_res</span><span class="p">,</span> <span class="n">mod_pos</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">mods</span>


<div class="viewcode-block" id="verify_location"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.verify_location">[docs]</a><span class="k">def</span> <span class="nf">verify_location</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">residue</span><span class="p">,</span> <span class="n">location</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the residue is at the given location in the UP sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein whose sequence is used as reference.</span>
<span class="sd">    residue : str</span>
<span class="sd">        A single character amino acid symbol (Y, S, T, V, etc.)</span>
<span class="sd">    location : str</span>
<span class="sd">        The location on the protein sequence (starting at 1) at which the</span>
<span class="sd">        residue should be checked against the reference sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the given residue is at the given position in the sequence</span>
<span class="sd">    corresponding to the given UniProt ID, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">get_sequence</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="c1"># If we couldn&#39;t get the sequence (can happen due to web service hiccups)</span>
    <span class="c1"># don&#39;t throw the statement away by default</span>
    <span class="k">if</span> <span class="n">seq</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">loc_int</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">location</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Invalid location </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">location</span><span class="p">)</span>
        <span class="n">loc_int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">loc_int</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">loc_int</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)):</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">elif</span> <span class="n">seq</span><span class="p">[</span><span class="n">loc_int</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">residue</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span></div>
    <span class="k">return</span> <span class="kc">False</span>


<div class="viewcode-block" id="verify_modification"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.verify_modification">[docs]</a><span class="k">def</span> <span class="nf">verify_modification</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">residue</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the residue at the given location has a known modifiation. </span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein whose sequence is used as reference.</span>
<span class="sd">    residue : str</span>
<span class="sd">        A single character amino acid symbol (Y, S, T, V, etc.)</span>
<span class="sd">    location : Optional[str]</span>
<span class="sd">        The location on the protein sequence (starting at 1) at which the</span>
<span class="sd">        modification is checked.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the given residue is reported to be modified at the given position </span>
<span class="sd">    in the sequence corresponding to the given UniProt ID, otherwise False.</span>
<span class="sd">    If location is not given, we only check if there is any residue of the</span>
<span class="sd">    given type that is modified.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mods</span> <span class="o">=</span> <span class="n">get_modifications</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="n">mod_locs</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mods</span><span class="p">]</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">get_sequence</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">location</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">verify_location</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">residue</span><span class="p">,</span> <span class="n">location</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">mod_idx</span> <span class="o">=</span> <span class="n">mod_locs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">location</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ml</span> <span class="ow">in</span> <span class="n">mod_locs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="n">ml</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">residue</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">True</span></div>
        <span class="k">return</span> <span class="kc">False</span>

<span class="k">def</span> <span class="nf">_is_organism</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">organism_suffix</span><span class="p">):</span>
    <span class="n">mnemonic</span> <span class="o">=</span> <span class="n">get_mnemonic</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mnemonic</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">mnemonic</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">organism_suffix</span><span class="p">):</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="kc">False</span>

<div class="viewcode-block" id="is_human"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.is_human">[docs]</a><span class="k">def</span> <span class="nf">is_human</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the given protein id corresponds to a human protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the protein_id corresponds to a human protein, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">_is_organism</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="s1">&#39;HUMAN&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="is_mouse"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.is_mouse">[docs]</a><span class="k">def</span> <span class="nf">is_mouse</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the given protein id corresponds to a mouse protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the protein_id corresponds to a mouse protein, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">_is_organism</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="s1">&#39;MOUSE&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="is_rat"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.is_rat">[docs]</a><span class="k">def</span> <span class="nf">is_rat</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the given protein id corresponds to a rat protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the protein_id corresponds to a rat protein, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">_is_organism</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="s1">&#39;RAT&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="get_mgi_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_mgi_id">[docs]</a><span class="k">def</span> <span class="nf">get_mgi_id</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the MGI ID given the protein id of a mouse protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the mouse protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mgi_id : str</span>
<span class="sd">        MGI ID of the mouse protein</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_mgi</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>

<div class="viewcode-block" id="get_rgd_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_rgd_id">[docs]</a><span class="k">def</span> <span class="nf">get_rgd_id</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the RGD ID given the protein id of a rat protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the rat protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    rgd_id : str</span>
<span class="sd">        RGD ID of the rat protein</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_rgd</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>

<div class="viewcode-block" id="get_id_from_mgi"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_id_from_mgi">[docs]</a><span class="k">def</span> <span class="nf">get_id_from_mgi</span><span class="p">(</span><span class="n">mgi_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt ID given the MGI ID of a mouse protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    mgi_id : str</span>
<span class="sd">        The MGI ID of the mouse protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    up_id : str</span>
<span class="sd">        The UniProt ID of the mouse protein.</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_mgi_reverse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mgi_id</span><span class="p">)</span>

<div class="viewcode-block" id="get_id_from_rgd"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_id_from_rgd">[docs]</a><span class="k">def</span> <span class="nf">get_id_from_rgd</span><span class="p">(</span><span class="n">rgd_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt ID given the RGD ID of a rat protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    rgd_id : str</span>
<span class="sd">        The RGD ID of the rat protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    up_id : str</span>
<span class="sd">        The UniProt ID of the rat protein.</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_rgd_reverse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rgd_id</span><span class="p">)</span>

<div class="viewcode-block" id="get_mouse_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_mouse_id">[docs]</a><span class="k">def</span> <span class="nf">get_mouse_id</span><span class="p">(</span><span class="n">human_protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the mouse UniProt ID given a human UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    human_protein_id : str</span>
<span class="sd">        The UniProt ID of a human protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mouse_protein_id : str</span>
<span class="sd">        The UniProt ID of a mouse protein orthologous to the given human protein</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_human_mouse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">human_protein_id</span><span class="p">)</span>

<div class="viewcode-block" id="get_rat_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_rat_id">[docs]</a><span class="k">def</span> <span class="nf">get_rat_id</span><span class="p">(</span><span class="n">human_protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the rat UniProt ID given a human UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    human_protein_id : str</span>
<span class="sd">        The UniProt ID of a human protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    rat_protein_id : str</span>
<span class="sd">        The UniProt ID of a rat protein orthologous to the given human protein</span>
<span class="sd">    &quot;&quot;&quot;</span></div>
    <span class="k">return</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_human_rat</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">human_protein_id</span><span class="p">)</span>


<div class="viewcode-block" id="get_function"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_function">[docs]</a><span class="k">def</span> <span class="nf">get_function</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the function description of a given protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        The UniProt ID of the protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        The function description of the protein.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">url</span> <span class="o">=</span> <span class="n">uniprot_url</span> <span class="o">+</span> <span class="n">protein_id</span> <span class="o">+</span> <span class="s1">&#39;.xml&#39;</span>
    <span class="n">ret</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="n">et</span> <span class="o">=</span> <span class="n">ElementTree</span><span class="o">.</span><span class="n">fromstring</span><span class="p">(</span><span class="n">ret</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>
    <span class="n">function</span> <span class="o">=</span> <span class="n">et</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;up:entry/up:comment[@type=&quot;function&quot;]/up:text&#39;</span><span class="p">,</span>
                       <span class="n">namespaces</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;up&#39;</span><span class="p">:</span> <span class="s1">&#39;http://uniprot.org/uniprot&#39;</span><span class="p">})</span>
    <span class="k">if</span> <span class="n">function</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span></div>
    <span class="k">return</span> <span class="n">function</span><span class="o">.</span><span class="n">text</span>


<span class="k">class</span> <span class="nc">UniprotMapper</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initialized_hmr</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">initialize</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">maps</span> <span class="o">=</span> <span class="n">_build_uniprot_entries</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_gene_name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_mnemonic</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_mnemonic_reverse</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_mgi</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_rgd</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_mgi_reverse</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_rgd_reverse</span> <span class="o">=</span> <span class="n">maps</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_sec</span> <span class="o">=</span> <span class="n">_build_uniprot_sec</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">def</span> <span class="nf">initialize_hmr</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_human_mouse</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_human_rat</span> <span class="o">=</span> \
            <span class="n">_build_human_mouse_rat</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initialized_hmr</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_gene_name</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_gene_name</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_mnemonic</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_mnemonic</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_mnemonic_reverse</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_mnemonic_reverse</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_mgi</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_mgi</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_rgd</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_rgd</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_mgi_reverse</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_mgi_reverse</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_rgd_reverse</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_rgd_reverse</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_sec</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_sec</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_human_mouse</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized_hmr</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize_hmr</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_human_mouse</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">uniprot_human_rat</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">initialized_hmr</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">initialize_hmr</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniprot_human_rat</span>

<span class="n">um</span> <span class="o">=</span> <span class="n">UniprotMapper</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">_build_uniprot_entries</span><span class="p">(</span><span class="n">from_pickle</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="c1"># Try the pickle file first if it exists (not in version control)</span>
    <span class="n">up_entries_pkl</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span> \
        <span class="s1">&#39;/../resources/uniprot_entries.pkl&#39;</span>
    <span class="k">if</span> <span class="n">from_pickle</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">up_entries_pkl</span><span class="p">):</span>
        <span class="kn">import</span> <span class="nn">pickle</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">up_entries_pkl</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="n">entries</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">entries</span>
    <span class="c1"># Otherwise process the regular entries TSV</span>
    <span class="n">up_entries_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span> \
        <span class="s1">&#39;/../resources/uniprot_entries.tsv&#39;</span>
    <span class="n">uniprot_gene_name</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_mnemonic</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_mnemonic_reverse</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_mgi</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_rgd</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_mgi_reverse</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_rgd_reverse</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">csv_rows</span> <span class="o">=</span> <span class="n">read_unicode_csv</span><span class="p">(</span><span class="n">up_entries_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="c1"># Skip the header row</span>
        <span class="nb">next</span><span class="p">(</span><span class="n">csv_rows</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_rows</span><span class="p">:</span>
            <span class="n">up_id</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">up_mnemonic</span><span class="p">,</span> <span class="n">rgd</span><span class="p">,</span> <span class="n">mgi</span> <span class="o">=</span> <span class="n">row</span>
            <span class="n">uniprot_gene_name</span><span class="p">[</span><span class="n">up_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_name</span>
            <span class="n">uniprot_mnemonic</span><span class="p">[</span><span class="n">up_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">up_mnemonic</span>
            <span class="n">uniprot_mnemonic_reverse</span><span class="p">[</span><span class="n">up_mnemonic</span><span class="p">]</span> <span class="o">=</span> <span class="n">up_id</span>
            <span class="k">if</span> <span class="n">mgi</span><span class="p">:</span>
                <span class="n">mgi_ids</span> <span class="o">=</span> <span class="n">mgi</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">mgi_ids</span><span class="p">:</span>
                    <span class="n">uniprot_mgi</span><span class="p">[</span><span class="n">up_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">mgi_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">uniprot_mgi_reverse</span><span class="p">[</span><span class="n">mgi_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">up_id</span>
            <span class="k">if</span> <span class="n">rgd</span><span class="p">:</span>
                <span class="n">rgd_ids</span> <span class="o">=</span> <span class="n">rgd</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">rgd_ids</span><span class="p">:</span>
                    <span class="n">uniprot_rgd</span><span class="p">[</span><span class="n">up_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">rgd_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">uniprot_rgd_reverse</span><span class="p">[</span><span class="n">rgd_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">up_id</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">uniprot_gene_name</span><span class="p">,</span> <span class="n">uniprot_mnemonic</span><span class="p">,</span> <span class="n">uniprot_mnemonic_reverse</span><span class="p">,</span> \
            <span class="n">uniprot_mgi</span><span class="p">,</span> <span class="n">uniprot_rgd</span><span class="p">,</span> <span class="n">uniprot_mgi_reverse</span><span class="p">,</span> <span class="n">uniprot_rgd_reverse</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">_build_human_mouse_rat</span><span class="p">():</span>
    <span class="n">hgnc_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span>\
                <span class="s1">&#39;/../resources/hgnc_entries.tsv&#39;</span>
    <span class="n">csv_rows</span> <span class="o">=</span> <span class="n">read_unicode_csv</span><span class="p">(</span><span class="n">hgnc_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="c1"># Skip the header row</span>
    <span class="nb">next</span><span class="p">(</span><span class="n">csv_rows</span><span class="p">)</span>
    <span class="n">uniprot_mouse</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_rat</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_rows</span><span class="p">:</span>
        <span class="n">human_id</span><span class="p">,</span> <span class="n">mgi_id</span><span class="p">,</span> <span class="n">rgd_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">6</span><span class="p">:</span><span class="mi">9</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">human_id</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">mgi_id</span><span class="p">:</span>
                <span class="n">mgi_id</span> <span class="o">=</span> <span class="n">mgi_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">mgi_id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;MGI:&#39;</span><span class="p">):</span>
                    <span class="n">mgi_id</span> <span class="o">=</span> <span class="n">mgi_id</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>
                <span class="n">mouse_id</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_mgi_reverse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mgi_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">mouse_id</span><span class="p">:</span>
                    <span class="n">uniprot_mouse</span><span class="p">[</span><span class="n">human_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">mouse_id</span>
            <span class="k">if</span> <span class="n">rgd_id</span><span class="p">:</span>
                <span class="n">rgd_id</span> <span class="o">=</span> <span class="n">rgd_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">rgd_id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;RGD:&#39;</span><span class="p">):</span>
                    <span class="n">rgd_id</span> <span class="o">=</span> <span class="n">rgd_id</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>
                <span class="n">rat_id</span> <span class="o">=</span> <span class="n">um</span><span class="o">.</span><span class="n">uniprot_rgd_reverse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rgd_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">rat_id</span><span class="p">:</span>
                    <span class="n">uniprot_rat</span><span class="p">[</span><span class="n">human_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">rat_id</span>
    <span class="k">return</span> <span class="n">uniprot_mouse</span><span class="p">,</span> <span class="n">uniprot_rat</span>

<span class="k">def</span> <span class="nf">_build_uniprot_sec</span><span class="p">():</span>
    <span class="c1"># Try loading pickle first (not in version control)</span>
    <span class="n">up_entries_pkl</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span> \
        <span class="s1">&#39;/../resources/uniprot_sec_ac.pkl&#39;</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">up_entries_pkl</span><span class="p">):</span>
        <span class="kn">import</span> <span class="nn">pickle</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">up_entries_pkl</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="n">entries</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">entries</span>
    <span class="c1"># File containing secondary accession numbers mapped</span>
    <span class="c1"># to primary accession numbers</span>
    <span class="n">sec_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span>\
                <span class="s1">&#39;/../resources/uniprot_sec_ac.txt&#39;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">uniprot_sec</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sec_file</span><span class="p">,</span> <span class="s1">&#39;rt&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lines</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">l</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;Secondary AC&#39;</span><span class="p">):</span>
                <span class="n">entry_lines</span> <span class="o">=</span> <span class="n">lines</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">:]</span>

        <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">entry_lines</span><span class="p">:</span>
            <span class="n">sec_id</span><span class="p">,</span> <span class="n">prim_id</span> <span class="o">=</span> <span class="n">l</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">uniprot_sec</span><span class="p">[</span><span class="n">sec_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prim_id</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">uniprot_sec</span><span class="p">[</span><span class="n">sec_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">prim_id</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">uniprot_sec</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">return</span> <span class="n">uniprot_sec</span>

<span class="k">def</span> <span class="nf">_build_uniprot_subcell_loc</span><span class="p">():</span>
    <span class="n">fname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span>\
                <span class="s1">&#39;/../resources/uniprot_subcell_loc.tsv&#39;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">csv_rows</span> <span class="o">=</span> <span class="n">read_unicode_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="c1"># Skip the header row</span>
        <span class="nb">next</span><span class="p">(</span><span class="n">csv_rows</span><span class="p">)</span>
        <span class="n">subcell_loc</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_rows</span><span class="p">:</span>
            <span class="n">loc_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">loc_alias</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">subcell_loc</span><span class="p">[</span><span class="n">loc_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">loc_alias</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">subcell_loc</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">return</span> <span class="n">subcell_loc</span>


<span class="n">uniprot_subcell_loc</span> <span class="o">=</span> <span class="n">_build_uniprot_subcell_loc</span><span class="p">()</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, B. M. Gyori, J. A. Bachman.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'1.10.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>